pyscf/pbc/df/fft_occk.py

```diff
@@ -0,0 +1,269 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Author: Peng Bao <baopeng@iccas.ac.cn>
+#
+# Modified from pyscf/pbc/df/fft_jk.py
+#
+# ref: Lin, L. J. Chem. Theory Comput. 2016, 12, 2242\u22122249.
+#
+
+'''
+JK with occ discrete Fourier transformation
+'''
+
+import time
+import numpy as np
+from pyscf import lib
+from pyscf.pbc import tools
+from pyscf.pbc.dft import numint
+from pyscf.pbc.df.df_jk import _format_dms, _format_kpts_band, _format_jks
+from pyscf.pbc.df.df_jk import _ewald_exxdiv_for_G0
+from pyscf.pbc.lib.kpts_helper import is_zero, gamma_point
+
+import scipy.linalg
+
+#@profile
+def get_k_kpts_occ(mydf, dm_kpts, hermi=1, kpts=np.zeros((1,3)), kpts_band=None,
+               exxdiv=None):
+    '''Get the Coulomb (J) and exchange (K) AO matrices at sampled k-points.
+
+    Args:
+        dm_kpts : (nkpts, nao, nao) ndarray
+            Density matrix at each k-point
+        kpts : (nkpts, 3) ndarray
+
+    Kwargs:
+        hermi : int
+            Whether K matrix is hermitian
+
+            | 0 : not hermitian and not symmetric
+            | 1 : hermitian
+
+        kpts_band : (3,) ndarray or (*,3) ndarray
+            A list of arbitrary "band" k-points at which to evalute the matrix.
+
+    Returns:
+        vj : (nkpts, nao, nao) ndarray
+        vk : (nkpts, nao, nao) ndarray
+        or list of vj and vk if the input dm_kpts is a list of DMs
+    '''
+    Jtime=time.time()
+
+    cell = mydf.cell
+    mesh = mydf.mesh
+    coords = cell.gen_uniform_grids(mesh)
+    ngrids = coords.shape[0]
+
+    if getattr(dm_kpts, 'mo_coeff', None) is not None:
+        mo_coeff = dm_kpts.mo_coeff
+        mo_occ   = dm_kpts.mo_occ
+    else:
+        mo_coeff = None
+
+    kpts = np.asarray(kpts)
+    dm_kpts = lib.asarray(dm_kpts, order='C')
+    dms = _format_dms(dm_kpts, kpts)
+    nset, nkpts, nao = dms.shape[:3]
+
+    weight = 1./nkpts * (cell.vol/ngrids)
+
+    kpts_band, input_band = _format_kpts_band(kpts_band, kpts), kpts_band
+    nband = len(kpts_band)
+
+    if gamma_point(kpts_band) and gamma_point(kpts):
+        vk_kpts = np.zeros((nset,nband,nao,nao), dtype=dms.dtype)
+    else:
+        vk_kpts = np.zeros((nset,nband,nao,nao), dtype=np.complex128)
+
+
+    coords = mydf.grids.coords
+    ao2_kpts = [np.asarray(ao.T, order='C')
+                for ao in mydf._numint.eval_ao(cell, coords, kpts=kpts)]
+    if input_band is None:
+        ao1_kpts = ao2_kpts
+    else:
+        ao1_kpts = [np.asarray(ao.T, order='C')
+                    for ao in mydf._numint.eval_ao(cell, coords, kpts=kpts_band)]
+
+######################## 
+    if mo_coeff is None:
+            return _format_jks(vk_kpts, dm_kpts, input_band, kpts)
```

Comment:
This will return a zero V_k at the 1st SCF iteration or if `dm_kpts` is not  properly tagged with the MO coefficients. It's dangerous to code it this way.